### ISSUE TITLE
Cold start reccomendations for unknown users (confirming approach)

### ISSUE THREAD

**Issue by nialloh23** [tensorflow/recommenders#124]
I'm trying to implement a retrieval model (e.g like the [context_features_tutorial](https://www.tensorflow.org/recommenders/examples/context_features)) that will handle cold start user predictions for users who are unknown to our model:

**Example query structure:**
```
ratings = ratings.map(lambda x: {
    "movie_title": x["movie_title"],
    "user_id": x["user_id"],
    "timestamp": x["timestamp"],
    "age": x["age"],
})
```

In the cold start scenario we will often have certain data missing (e.g. user_id, age).

From the tutorials it would appear that the following should work:

1. Pass the unknown user_id to the index -> the StringLookup layer will determine that this user is not in our vocabulary and will assign it to the `OOV / UNK` embedding. My assumption is this will also work the same for any other unknown categorical vars handled via StringLookUp.

2. For missing numerical features -> it would appear from ([issue 108](https://github.com/tensorflow/recommenders/issues/108)) that the discretization layer cannot handle missing or None values. The suggested approach in issue 108 is to first use IntegerLookup to encode missing values and then use the bin discretization layer afterwards. One other approach we're considering is just replacing all missing numerical features wiht their respective means before feeding them to the model (e.g. mean age, mean_time_watched)

My question are:
1. Is this the correct way to approach making predictions for cold start users?
2. Does the 'OOV/UNK' treatment work the same way for user_ids not in the vocab (e.g. '1013413') and missing user ids ('') 

Thanks for any guidance on the above

**Comment by maciejkula** [comment]
Thanks for the detailed question!

Some general points first:
1. With cold-start users and items it will be beneficial to use any other features you have about them: location, device, description and so forth.
2. If the user is cold-start because you haven't retrained your model yet, sequence-based models that take past interactions to compute the user's representations are really useful (see [this](https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/46488.pdf) or the GRU4Rec paper).

On your approach:
1. Yes, that's correct. By the way, you can set the number of OOV indices to more than one ([docs](https://www.tensorflow.org/api_docs/python/tf/keras/layers/experimental/preprocessing/IntegerLookup)). This will apply the hashing trick to any entries not in the vocabulary.
2. I'm afraid I don't have good suggestions for treating missing continuous variables. If you're discretizing them, it definitely makes sense to treat missing values as their own bin; if you're keeping it continuous mean imputation might be OK.

As for the specific treatment of ids not in the vocab/empty string: I'd just open up a Python shell and see :). As far as the model is concerned, though, these are the same: they are just different possible values for a categorical variable.

**Comment by nialloh23** [comment]
Thanks Maciej. That all makes sense.

**Comment by cilista** [comment]
Hi, I'm picking up this thread after a long time to clarify a doubt. Like @nialloh23 , I'm developing a retrieval model with a two-tower approach using both user and product features.

My doubt is about new users:

Obviously, during training I don't have OOV/UKN users, only during inference.
If we never have OOV/UKN during training, how will they be placed in the embedding space?
Would it be like using only the information available at inference time and "ignoring" the user_id?
I hope I was clear.

Thanks a lot!

**Comment by patrickorlando** [comment]
You can randomly mask the `user_id` for new users during training @cilista.
If possible you should also aim to have generalisable new user features.

**Comment by cilista** [comment]
Hi @patrickorlando , first thanks a lot for your reply.

At inference time, I have all the information related to the user except for the user_id, while during training the user_id is always present.

If I got it right, the idea is to mask the user_id for a percentage of users during training and use it as "OOV", so that it can be used as a sort of "average user" in the embedding space?

**Comment by patrickorlando** [comment]
You should simulate the scenario you encounter at inference time. If retraining occurs daily then a user_id will be `[UNK]` for the user's first day. Therefor, you should randomly set the user_id to `[UNK]` with some probability `p`, but only for records that are the first 24 hours for that user.

**Comment by cilista** [comment]
@patrickorlando you made my day!
